File: backend/app/services/kpis.py
from sqlalchemy.orm import Session
from sqlalchemy import func, case, distinct, extract, desc, text
from datetime import datetime, timedelta
from typing import List, Literal
from ..models import Prompt, UsageEvent, Deployment, ABPolicy, ABAssignment


def get_summary(db: Session) -> dict:
    """
    Calculate summary metrics for dashboard cards:
    - total prompts (distinct names)
    - active prompts (distinct names where active=True)
    - total deployments
    - running experiments (active ABPolicies)
    - usage in last 7 days
    """

    # Total prompts (distinct names)
    total_prompts = db.query(func.count(distinct(Prompt.name))).scalar() or 0

    # Active prompts: prompts where latest version is active=True
    # Subquery to get max version per name
    max_versions = (
        db.query(
            Prompt.name,
            func.max(Prompt.version).label("max_version")
        )
        .group_by(Prompt.name)
        .subquery()
    )

    active_prompts = (
        db.query(func.count(distinct(Prompt.name)))
        .join(
            max_versions,
            (Prompt.name == max_versions.c.name) &
            (Prompt.version == max_versions.c.max_version)
        )
        .filter(Prompt.active == True)
        .scalar() or 0
    )

    # Total deployments
    total_deployments = db.query(func.count(Deployment.id)).scalar() or 0

    # Running experiments (count ABPolicy records)
    running_experiments = db.query(func.count(ABPolicy.id)).scalar() or 0

    # Usage in last 7 days
    seven_days_ago = datetime.utcnow() - timedelta(days=7)
    usage_7d = (
        db.query(func.count(UsageEvent.id))
        .filter(UsageEvent.created_at >= seven_days_ago)
        .scalar() or 0
    )

    return {
        "prompts": total_prompts,
        "active_prompts": active_prompts,
        "deployments": total_deployments,
        "experiments": running_experiments,
        "usage_7d": usage_7d,
    }


def get_usage_trend(
    db: Session,
    period_days: int = 42,
    bucket: Literal["week", "day"] = "week"
) -> List[dict]:
    """
    Get time series of usage metrics grouped by week or day.
    Returns list of data points with executions, failures, avg_latency, avg_cost.
    """

    cutoff_date = datetime.utcnow() - timedelta(days=period_days)

    # PostgreSQL date_trunc function
    if bucket == "week":
        trunc_expr = func.date_trunc("week", UsageEvent.created_at)
    else:
        trunc_expr = func.date_trunc("day", UsageEvent.created_at)

    results = (
        db.query(
            trunc_expr.label("bucket_start"),
            func.count(UsageEvent.id).label("executions"),
            func.sum(case((UsageEvent.success == False, 1), else_=0)).label("failures"),
            func.avg(UsageEvent.latency_ms).label("avg_latency"),
            func.avg(UsageEvent.cost).label("avg_cost"),
        )
        .filter(UsageEvent.created_at >= cutoff_date)
        .group_by("bucket_start")
        .order_by("bucket_start")
        .all()
    )

    points = []
    for row in results:
        points.append({
            "start": row.bucket_start.strftime("%Y-%m-%d"),
            "executions": int(row.executions or 0),
            "failures": int(row.failures or 0),
            "avg_latency": float(row.avg_latency) if row.avg_latency is not None else None,
            "avg_cost": float(row.avg_cost) if row.avg_cost is not None else None,
        })

    return points


def get_version_velocity(db: Session, months: int = 6) -> List[dict]:
    """
    Count version releases per month for the last N months.
    Returns list of {month: "YYYY-MM", releases: int}
    """

    cutoff_date = datetime.utcnow() - timedelta(days=months * 30)

    results = (
        db.query(
            func.date_trunc("month", Prompt.created_at).label("month"),
            func.count(Prompt.id).label("releases"),
        )
        .filter(Prompt.created_at >= cutoff_date)
        .group_by("month")
        .order_by("month")
        .all()
    )

    points = []
    for row in results:
        points.append({
            "month": row.month.strftime("%Y-%m"),
            "releases": int(row.releases or 0),
        })

    return points


def get_top_prompts(db: Session, limit: int = 10, period_days: int = 30) -> List[dict]:
    """
    Get top N prompts by execution volume in the last period_days.
    Includes: name, executions, success_rate, avg_cost, last_updated
    """

    cutoff_date = datetime.utcnow() - timedelta(days=period_days)

    # Join UsageEvent with Prompt to get prompt names
    # Group by prompt name, calculate metrics
    results = (
        db.query(
            Prompt.name,
            func.count(UsageEvent.id).label("executions"),
            (
                func.sum(case((UsageEvent.success == True, 1), else_=0)) * 1.0 /
                func.nullif(func.count(UsageEvent.id), 0)
            ).label("success_rate"),
            func.avg(UsageEvent.cost).label("avg_cost"),
            func.max(Prompt.created_at).label("last_updated"),
        )
        .join(Prompt, Prompt.id == UsageEvent.prompt_id)
        .filter(UsageEvent.created_at >= cutoff_date)
        .group_by(Prompt.name)
        .order_by(desc("executions"))
        .limit(limit)
        .all()
    )

    items = []
    for row in results:
        items.append({
            "name": row.name,
            "executions": int(row.executions or 0),
            "success_rate": float(row.success_rate or 0.0),
            "avg_cost": float(row.avg_cost) if row.avg_cost is not None else None,
            "last_updated": row.last_updated,
        })

    return items


def get_experiments(db: Session) -> List[dict]:
    """
    Get summary of active A/B experiments.
    For each ABPolicy, return arms with version, weight, assignments, success_rate.
    """

    # Get all active policies
    policies = db.query(ABPolicy).all()

    experiments = []

    for policy in policies:
        arms = []

        # Parse weights from policy (e.g., {"1": 50, "2": 50})
        weights = policy.weights or {}

        for version_str, weight in weights.items():
            version = int(version_str)

            # Count assignments for this version
            assignment_count = (
                db.query(func.count(ABAssignment.id))
                .filter(
                    ABAssignment.prompt_name == policy.prompt_name,
                    ABAssignment.version == version
                )
                .scalar() or 0
            )

            # Calculate success rate for this version
            # Find prompt_id for this version
            prompt = (
                db.query(Prompt)
                .filter(Prompt.name == policy.prompt_name, Prompt.version == version)
                .first()
            )

            success_rate = None
            if prompt:
                stats = (
                    db.query(
                        func.sum(case((UsageEvent.success == True, 1), else_=0)).label("successes"),
                        func.count(UsageEvent.id).label("total"),
                    )
                    .filter(UsageEvent.prompt_id == prompt.id)
                    .first()
                )

                if stats and stats.total and stats.total > 0:
                    success_rate = float(stats.successes or 0) / float(stats.total)

            arms.append({
                "version": version,
                "weight": float(weight) / 100.0,  # Convert to decimal (0.5 instead of 50)
                "assignments": assignment_count,
                "success_rate": success_rate,
            })

        # Use prompt_name as experiment name (or could derive from ABAssignment if needed)
        experiments.append({
            "experiment": policy.prompt_name,  # Can be enhanced with experiment_name from ABAssignment
            "prompt": policy.prompt_name,
            "arms": arms,
        })

    return experiments

File: frontend/src/lib/api/kpis.ts
/**
 * KPIs API Functions
 *
 * Functional API for KPI and analytics operations.
 * Designed for use with TanStack Query.
 */

import { createClient } from '@/utils/supabase/client';

export interface SummaryMetrics {
  total_prompts: number;
  active_prompts: number;
  total_deployments: number;
  running_experiments: number;
  recent_usage: number;
}

export interface UsageTrendData {
  period: string;
  executions: number;
  failures: number;
  avg_latency: number | null;
  avg_cost: number | null;
}

export interface VersionVelocityData {
  month: string;
  releases: number;
}

export interface TopPromptData {
  name: string;
  executions: number;
  success_rate: number;
  avg_cost: number | null;
  last_updated: string;
}

export interface ExperimentAnalytics {
  experiment: string;
  prompt: string;
  arms: Array<{
    version: number;
    weight: number;
    assignments: number;
    success_rate: number | null;
  }>;
}

/**
 * Get summary metrics for dashboard
 */
export async function get_summary(): Promise<SummaryMetrics> {
  const supabase = createClient();

  // Get total prompts (distinct names)
  const { count: total_prompts } = await (
    supabase.from('prompts') as any
  ).select('name', { count: 'exact', head: true });

  // Get active prompts (latest version active)
  const { data: all_prompts } = await (supabase.from('prompts') as any)
    .select('name, version, active')
    .order('version', { ascending: false });

  const active_prompts_set = new Set();
  const seen_names = new Set();

  (all_prompts || []).forEach(
    (p: { name: string; version: number; active: boolean }) => {
      if (!seen_names.has(p.name)) {
        seen_names.add(p.name);
        if (p.active) {
          active_prompts_set.add(p.name);
        }
      }
    }
  );

  // Get total deployments
  const { count: total_deployments } = await (
    supabase.from('deployments') as any
  ).select('*', { count: 'exact', head: true });

  // Get running experiments (count of AB policies)
  const { count: running_experiments } = await (
    supabase.from('ab_policies') as any
  ).select('*', { count: 'exact', head: true });

  // Get usage in last 7 days
  const seven_days_ago = new Date();
  seven_days_ago.setDate(seven_days_ago.getDate() - 7);

  const { count: recent_usage } = await (
    supabase.from('usage_events') as any
  )
    .select('*', { count: 'exact', head: true })
    .gte('created_at', seven_days_ago.toISOString());

  return {
    total_prompts: total_prompts || 0,
    active_prompts: active_prompts_set.size,
    total_deployments: total_deployments || 0,
    running_experiments: running_experiments || 0,
    recent_usage: recent_usage || 0,
  };
}

/**
 * Get usage trend over time
 */
export async function get_usage_trend(
  period_days = 42,
  bucket: 'week' | 'day' = 'week'
): Promise<UsageTrendData[]> {
  const supabase = createClient();

  const start_date = new Date();
  start_date.setDate(start_date.getDate() - period_days);

  const { data: events } = await (supabase.from('usage_events') as any)
    .select('created_at, success, latency_ms, cost')
    .gte('created_at', start_date.toISOString())
    .order('created_at', { ascending: true });

  if (!events || events.length === 0) {
    return [];
  }

  // Group by time bucket
  const buckets = new Map<
    string,
    {
      executions: number;
      failures: number;
      total_latency: number;
      total_cost: number;
    }
  >();

  events.forEach((event: any) => {
    const date = new Date(event.created_at);
    let bucket_key: string;

    if (bucket === 'week') {
      // Get week start (Monday)
      const day_of_week = date.getDay();
      const diff = date.getDate() - day_of_week + (day_of_week === 0 ? -6 : 1);
      const week_start = new Date(date.setDate(diff));
      bucket_key = week_start.toISOString().split('T')[0];
    } else {
      bucket_key = date.toISOString().split('T')[0];
    }

    if (!buckets.has(bucket_key)) {
      buckets.set(bucket_key, {
        executions: 0,
        failures: 0,
        total_latency: 0,
        total_cost: 0,
      });
    }

    const stats = buckets.get(bucket_key)!;
    stats.executions++;
    if (!event.success) stats.failures++;
    if (event.latency_ms) stats.total_latency += event.latency_ms;
    if (event.cost) stats.total_cost += event.cost;
  });

  // Convert to array
  return Array.from(buckets.entries()).map(([period, stats]) => ({
    period,
    executions: stats.executions,
    failures: stats.failures,
    avg_latency:
      stats.executions > 0 ? stats.total_latency / stats.executions : null,
    avg_cost:
      stats.executions > 0 ? stats.total_cost / stats.executions : null,
  }));
}

/**
 * Get version release velocity
 */
export async function get_version_velocity(
  months = 6
): Promise<VersionVelocityData[]> {
  const supabase = createClient();

  const start_date = new Date();
  start_date.setMonth(start_date.getMonth() - months);

  const { data: prompts } = await (supabase.from('prompts') as any)
    .select('created_at')
    .gte('created_at', start_date.toISOString())
    .order('created_at', { ascending: true });

  if (!prompts || prompts.length === 0) {
    return [];
  }

  // Group by month
  const month_counts = new Map<string, number>();

  prompts.forEach((prompt: any) => {
    const date = new Date(prompt.created_at);
    const month_key = `${date.getFullYear()}-${String(date.getMonth() + 1).padStart(2, '0')}`;

    month_counts.set(month_key, (month_counts.get(month_key) || 0) + 1);
  });

  return Array.from(month_counts.entries()).map(([month, count]) => ({
    month,
    releases: count,
  }));
}

/**
 * Get top prompts by usage
 */
export async function get_top_prompts(
  limit = 10,
  period_days = 30
): Promise<TopPromptData[]> {
  const supabase = createClient();

  const start_date = new Date();
  start_date.setDate(start_date.getDate() - period_days);

  // Get usage events in the period
  const { data: events } = await (supabase.from('usage_events') as any)
    .select('prompt_id, success, cost')
    .gte('created_at', start_date.toISOString());

  if (!events || events.length === 0) {
    return [];
  }

  // Group by prompt_id
  const prompt_stats = events.reduce(
    (
      acc: Record<
        number,
        { executions: number; success_count: number; total_cost: number }
      >,
      event: any
    ) => {
      if (!acc[event.prompt_id]) {
        acc[event.prompt_id] = {
          executions: 0,
          success_count: 0,
          total_cost: 0,
        };
      }

      acc[event.prompt_id].executions++;
      if (event.success) acc[event.prompt_id].success_count++;
      if (event.cost) acc[event.prompt_id].total_cost += event.cost;

      return acc;
    },
    {} as Record<
      number,
      { executions: number; success_count: number; total_cost: number }
    >
  );

  // Get prompt details
  const prompt_ids = Object.keys(prompt_stats).map(Number);

  const { data: prompts } = await (supabase.from('prompts') as any)
    .select('id, name, version, created_at')
    .in('id', prompt_ids);

  if (!prompts) {
    return [];
  }

  // Combine and sort
  const result = prompts.map((prompt: any) => {
    const stats = prompt_stats[prompt.id];
    return {
      name: prompt.name,
      executions: stats.executions,
      success_rate:
        stats.executions > 0 ? stats.success_count / stats.executions : 0,
      avg_cost:
        stats.executions > 0 ? stats.total_cost / stats.executions : null,
      last_updated: prompt.created_at,
    };
  });

  // Sort by executions and take top N
  result.sort((a: any, b: any) => b.executions - a.executions);

  return result.slice(0, limit);
}

/**
 * Get active experiments summary
 */
export async function get_experiments_analytics(): Promise<
  ExperimentAnalytics[]
> {
  const supabase = createClient();

  const { data: policies } = await (supabase.from('ab_policies') as any)
    .select('*')
    .eq('is_public', true);

  if (!policies || policies.length === 0) {
    return [];
  }

  // For each policy, get assignment stats
  const experiments = await Promise.all(
    policies.map(async (policy: any) => {
      const { data: assignments } = await (
        supabase.from('ab_assignments') as any
      )
        .select('*')
        .eq('prompt_name', policy.prompt_name);

      const weights = policy.weights as Record<string, number>;
      const versions = Object.keys(weights).map(Number);

      // Count assignments per version
      const version_counts = (assignments || []).reduce(
        (acc: Record<number, number>, a: any) => {
          acc[a.version] = (acc[a.version] || 0) + 1;
          return acc;
        },
        {} as Record<number, number>
      );

      const arms = versions.map((version) => ({
        version,
        weight: weights[version],
        assignments: version_counts[version] || 0,
        success_rate: null, // Would need usage_events joined by prompt version
      }));

      return {
        experiment: policy.prompt_name,
        prompt: policy.prompt_name,
        arms,
      };
    })
  );

  return experiments;
}
